model/src/main/kotlin/io/spine/validation/range/NumericBound.kt

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2025, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.validation.range
+
+import io.spine.protodata.Compilation
+import io.spine.protodata.ast.PrimitiveType
+import io.spine.protodata.ast.PrimitiveType.TYPE_DOUBLE
+import io.spine.protodata.ast.PrimitiveType.TYPE_FIXED32
+import io.spine.protodata.ast.PrimitiveType.TYPE_FIXED64
+import io.spine.protodata.ast.PrimitiveType.TYPE_FLOAT
+import io.spine.protodata.ast.PrimitiveType.TYPE_INT32
+import io.spine.protodata.ast.PrimitiveType.TYPE_INT64
+import io.spine.protodata.ast.PrimitiveType.TYPE_SFIXED32
+import io.spine.protodata.ast.PrimitiveType.TYPE_SFIXED64
+import io.spine.protodata.ast.PrimitiveType.TYPE_SINT32
+import io.spine.protodata.ast.PrimitiveType.TYPE_SINT64
+import io.spine.protodata.ast.PrimitiveType.TYPE_UINT32
+import io.spine.protodata.ast.PrimitiveType.TYPE_UINT64
+import io.spine.protodata.ast.name
+import io.spine.protodata.ast.qualifiedName
+import io.spine.protodata.check
+import io.spine.validation.RANGE
+import io.spine.validation.NumericBound as ProtoNumericBound
+
+/**
+ * One to one Kotlin representation of [ProtoNumericBound].
+ *
+ * We would like to have a Kotlin counterpart because of [UInt] and [ULong] types.
+ * It eases the implementation of parsing and comparisons for such bounds.
+ * Otherwise, we would have to care for special cases with unsigned types (which are just `int`
+ * and `long` in Protobuf) when parsing and comparing such bounds. With this class,
+ * Kotlin unsigned classes do it for us.
+ */
+internal data class NumericBound(
+    val value: Any, // Cannot use `Number` because Kotlin's `UInt` and `ULong` are not numbers.
+    val inclusive: Boolean
+) : Comparable<NumericBound> {
```

Comment:
Isn't it possible to make the Proto `NumericBound` implement some interface, which would extend `Comparable`, with the default implementation of `compareTo`?

Reply:
It is possible to create a `Mixin`, but with a Kotlin class we don't do the work twice.

Now we parse `string` to a Kotlin type, compare it with them and transform to Proto.

With `Mixin`, we parse `string` to a Kotlin type (to know it parses well), set the parsed value to a Proto message (throwing away already created Kotlin instance). And when comparing, create a Kotlin type from Proto for comparisons.